ds701_book/27-RNN.qmd

```diff
@@ -0,0 +1,640 @@
+hidden_dim = 20
+num_layers = 2
+output_dim = X.shape[2]
+model = MusicLSTM(input_dim, hidden_dim, num_layers, output_dim)
+```
+
+
+## Training
+
+We will train our model for 10 epochs.
+
+```{python}
+criterion = nn.MSELoss()
+optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
+
+num_epochs = 100
+batch_size = 8
+for epoch in range(num_epochs):
+    for i in range(0, len(X), batch_size):
+        inputs = torch.tensor(X[i:i+batch_size], dtype=torch.float32)
+        targets = torch.tensor(y[i:i+batch_size], dtype=torch.float32)
+        
+        outputs = model(inputs)
+        loss = criterion(outputs, targets)
+        
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+```
+
+## Generate
+
+We next write a function that will help us generate a new piece of music.
+
+```{python}
+#| code-fold: false
+def generate_music(model, start_sequence, length):
+    model.eval()
+    generated = start_sequence
+    for _ in range(length):
+        input_seq = torch.tensor(generated[-sequence_length:], dtype=torch.float32).unsqueeze(0)
+        next_note = model(input_seq).detach().numpy()
+        generated = np.vstack((generated, next_note))
+    return generated
+
+# Example usage
+start_sequence = X[10]  # Starting sequence for generation
+generated_music = generate_music(model, start_sequence, 100)  # Generate 100 notes
+print(generated_music.shape)
+```
+
+## Download the Song
+
+We can generate an output midi file to play. The song is somewhat similar to the original piece. We could probably create a better song through more training and a larger training set.
+
+```{python}
+def notes_to_midi(notes, output_file):
+    midi = pretty_midi.PrettyMIDI()
+    instrument = pretty_midi.Instrument(program=0)
+    for note in notes:
+        start, end, pitch, velocity = note
+        midi_note = pretty_midi.Note(
+            velocity=int(velocity),
+            pitch=int(pitch),
+            start=float(start),
+            end=float(end)
+        )
+        instrument.notes.append(midi_note)
+    midi.instruments.append(instrument)
+    midi.write(output_file)
+
+# Uncomment this line to write the follow
+#notes_to_midi(generated_music, 'generated_music.midi')
+```
+
+## Recap
+
+We discussed the basic RNN architecture. We then discussed the LSTM and GRU modifications. These modifications allowed the RNNs to handle long-term dependencies.
+
+We then considered two example applications. One of energy consumption prediction and one of of music generation. We generated a piano solo that sounded *interesting*.
+
+One computational bottleneck of RNNs is that they only allow for sequential computation. This was a major bottleneck for applications involving natural language processing. This motivated the development of more advanced language processing architectures, in particular, the transformer, which is the basis for modern large language models (LLMs).
```

Comment:
We should mention that the architecture evolved to add attention (cite paper), and then famously got rid of the RNN part and just kept attention and the linear layers in "Attention is all you need" and hence the Transformer. I still think RNNs can be useful however.

Reply:
Maybe we should allude to his in the intro material too. What do you think?

Reply:
I was thinking of this as more of setup for the next NLP lectures, but I can mention it in the beginning too.

Reply:
Do you want me to cite the attention is all you need paper? OR is there a different paper you had in mind?

Reply:
Is this the paper to cite: 

D. Bahdanau, K. Cho, and Y. Bengio, “Neural Machine Translation by Jointly Learning to Align and Translate.” arXiv, May 19, 2016. Accessed: Feb. 16, 2023. [Online]. Available: http://arxiv.org/abs/1409.0473
2014?

Reply:
yes.

Reply:
I added a What's Next slide at the end to include this information. I feel that adding all this in the beginning will require too much explanation and the segue into transformers and NLP makes more sense at the end.